mirgecom/artificial_viscosity.py

```diff
@@ -0,0 +1,346 @@
+    r_flux_db = 0
+    for btag in boundaries:
+        r_tpair = TracePair(
+            btag,
+            interior=discr.project("vol", btag, r),
+            exterior=boundaries[btag].exterior_grad_q(discr, btag=btag, t=t,
+                                                      grad_q=r, eos=eos))
+        r_flux_db = r_flux_db + _facial_flux_r(discr, r_tpair=r_tpair)
+    # Total grad(Q) flux element boundaries
+    r_flux_bnd = r_flux_int + r_flux_pb + r_flux_db
+
+    # Return the AV RHS term
+    return discr.inverse_mass(-div_r_vol + discr.face_mass(r_flux_bnd))
+
+
+def artificial_viscosity(discr, t, eos, boundaries, q, alpha, **kwargs):
+    """Interface :function:`av_operator` with backwards-compatible API."""
+    from warnings import warn
+    warn("Do not call artificial_viscosity; it is now called av_operator. This"
+         "function will disappear in 2021", DeprecationWarning, stacklevel=2)
+    return av_operator(discr=discr, eos=eos, boundaries=boundaries,
+                       q=q, alpha=alpha, t=t)
+
+
+def linear_operator_kernel():
+    """Apply linear operator to all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in}""",
+        "result[iel,idof] = sum(j, mat[idof, j] * vec[iel, j])",
+        name="modal_decomp",
+    )
+    knl = lp.tag_array_axes(knl, "mat", "stride:auto,stride:auto")
+    return knl
+
+
+def compute_smoothness_indicator():
+    """Compute the smoothness indicator for all elements."""
+    from meshmode.array_context import make_loopy_program
+
+    knl = make_loopy_program(
+        """{[iel,idof,j,k]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in and
+        0<=k<ndiscr_nodes_in}""",
+        "result[iel,idof] = "
+        "sum(k,vec[iel,k]*vec[iel,k]*modes[k])/sum(j,"
+        " vec[iel,j]*vec[iel,j]+1.0e-12/ndiscr_nodes_in)",
+        name="smooth_comp",
+    )
+    return knl
+
+
+def smoothness_indicator(discr, u, kappa=1.0, s0=-6.0):
+    """Calculate the smoothness indicator.
+
+    Parameters
+    ----------
+    u
+        A DOF Array of the field that is used to calculate the
+        smoothness indicator.
+
+    kappa
+        A optional argument that sets the controls the width of the
+        transition between 0 to 1.
+    s0
+        A optional argument that sets the smoothness level to limit
+        on. Logical values are [0,-infinity) where -infinity results in
+        all cells being tagged and 0 results in none.
+
+    Returns
+    -------
+    meshmode.dof_array.DOFArray
+        A DOF Array containing elementwise constant values between 0 and 1
+        which indicate the smoothness of a given element.
+    """
+    assert isinstance(u, DOFArray)
+
+    def get_kernel():
+        return linear_operator_kernel()
```

Comment:
This can go away with https://github.com/inducer/grudge/pull/77

Reply:
Resolved: https://github.com/illinois-ceesd/mirgecom/pull/213/commits/7dfa8143c5162dada86b5d819469fe53d42128ff